Photon: Fix PHP warning when Photonizing some URLs

bisko opened the pull request:
#### Changes proposed in this Pull Request:

* Fix an issue where some URLs would generate a PHP warning when trying to replace them with their Photon counterpart. 
* Fix failing `Jetpack_Photon` unit tests.

#### Jetpack product discussion

N/a

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

The quickest way to test is via `wp shell` on the local Jetpack instance.

If you have a local Jetpack Docker instance: 

* `yarn docker:wp shell --url=http://<your test site URL>`
* Then in the shell context run: 
   ```
   \Jetpack_Photon::filter_the_content("<a href=\"https://test.com/foo.jpg#d\" ><img src=\"http://test.com/8.jpg\"/></a>\n\n\n<a href=\"https://test.com/foo.jpg#d\" ><img src=\"http://i0.wp.com/8.jpg\"/></a>");
   ```
* In the output you should see two photonized image `src` and links (`a href`), similar to:
   ```
   <a href="https://i2.wp.com/test.com/foo.jpg?ssl=1" ><img src="https://i0.wp.com/test.com/8.jpg?w=580" data-recalc-dims="1"/></a>\n
   <a href="https://i2.wp.com/test.com/foo.jpg?ssl=1" ><img src="http://i0.wp.com/8.jpg"/></a>
   ```

If you don't have a local Docker instance:

* Enable Photon
* Try to publish a post with the following content:
   ```
   <a href="https://test.com/foo.jpg#d" ><img src="http://test.com/8.jpg"/></a>
   <a href="https://test.com/foo.jpg#d" ><img src="http://i0.wp.com/8.jpg"/></a>
   ```
* Make sure the output is similar to the Docker case above.

Unit Testing - if you have the local Docker Instance, as above:

```
yarn docker:phpunit --filter photon_
```

This will run all the `Jetpack_Photon` unit tests. They should all pass.

#### Proposed changelog entry for your changes:

* Fix PHP warning when replacing URL with their Photon counterpart

bisko committed:
Fix PHP warning in Photon

Fixes a PHP warning when trying to Photonize more exotic URLs

jeherve reviewed (changes requested):
Do you think you could add a test to cover this use-case in `tests/php/test_class.jetpack_photon.php`?

bisko committed:
Add Unit tests

bisko commented on tests/php/test_class.jetpack_photon.php:
```diff
@@ -229,6 +229,19 @@ public function test_photon_parse_images_from_html_a_tags_without_images() {
+	public function test_photon_parse_images_from_html_a_tags_with_hash_in_href() {
+		list( $sample_html, $expected ) = $this->get_photon_sample_content( 'a-tags-with-hash-href.html' );
+
+		$this->assertEquals( trim( $expected ), trim( Jetpack_Photon::filter_the_content( $sample_html ) ) );
```
`trim()` because IDE wants to add a new line at the end of the file, which gets added to the expected results :/

bisko commented:
> Do you think you could add a test to cover this use-case in `tests/php/test_class.jetpack_photon.php`?

Thanks for the nudge. Added.

jeherve commented:
I rebased to get rid of the unrelated test failures, but the tests are still failing wit the new addition:
https://travis-ci.com/github/Automattic/jetpack/jobs/383091048

bisko committed:
Try to fix Travis CI tests that differ from local instance

bisko committed:
Fix failing unit test `test_photon_parse_images_from_html_a_tags_with_hash_in_href` by resetting parameters that are autodetected in the `filter_the_content` method.

Fix most other Photon unit tests that were failing due to missing variable reset on `tearDown`.

bisko committed:
Try to please older PHP versions

bisko commented:
> I rebased to get rid of the unrelated test failures, but the tests are still failing wit the new addition:

Fixed both the failing test about this PR, as well as all the failing Photon tests. All should be passing correctly now.

jeherve commented:
Cherry-picked to `branch-8.9` in 0d8a0ae3e0a4cd5db5abe0bdd4899acfff8c582f
